hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/DirectStoreFSWriteStrategy.java

```diff
@@ -0,0 +1,258 @@
+   * @return
+   * @throws IOException
+   */
+  @Override
+  public Path splitStoreFile(RegionInfo parentRegionInfo, RegionInfo daughterRegionInfo,
+    String familyName, HStoreFile f, byte[] splitRow,
+    boolean top, RegionSplitPolicy splitPolicy, FileSystem fs) throws IOException {
+    Path path = super.splitStoreFile(parentRegionInfo, daughterRegionInfo,
+      familyName, f, splitRow, top, splitPolicy, fs);
+    Map<String,List<Path>> splitReferences = regionSplitReferences.
+      get(daughterRegionInfo.getEncodedName());
+    if(splitReferences==null){
+      splitReferences = new HashMap<>();
+      regionSplitReferences.put(daughterRegionInfo.getEncodedName(), splitReferences);
+    }
+    List<Path> references = splitReferences.get(familyName);
+    if(references==null){
+      references = new ArrayList<>();
+      splitReferences.put(familyName,references);
+    }
+    references.add(path);
+    return path;
+  }
+
+  /**
+   * Wraps <code>super.mergeStoreFile</code>, so that it can map the resulting merge reference to
+   * the related merged region and column family, in order to add this to 'storefile' system table
+   * for the tracking logic of PersisedStoreFileManager later in <code>commitMergedRegion</code>.
+   * @param regionToMerge {@link RegionInfo} for one of the regions being merged.
+   * @param mergedRegion {@link RegionInfo} of the merged region
+   * @param familyName Column Family Name
+   * @param f File to create reference.
+   * @param mergedDir
+   * @param fs FileSystem instance for creating the actual reference file.
+   * @return
+   * @throws IOException
+   */
+  @Override
+  public Path mergeStoreFile(RegionInfo regionToMerge, RegionInfo mergedRegion, String familyName,
+      HStoreFile f, Path mergedDir, FileSystem fs) throws IOException {
+//    Path path = super.mergeStoreFile(regionToMerge, mergedRegion, familyName, f, mergedDir, fs);
+
+    Path path = (this.fileSystem.regionInfoForFs.equals(regionToMerge)) ?
+      super.mergeStoreFile(mergedRegion, regionToMerge, familyName, f, mergedDir, fs)
+      : super.mergeStoreFile(regionToMerge, mergedRegion, familyName, f, mergedDir, fs);
+
+    List<Path> referenceFiles = mergeReferences.get(familyName);
+    if(referenceFiles==null){
+      referenceFiles = new ArrayList<>();
+      mergeReferences.put(familyName, referenceFiles);
+    }
+    referenceFiles.add(path);
+    return path;
+  }
+
+  /**
+   * Do nothing. Here the split dir is the table dir itself, we cannot delete it.
+   * @throws IOException
+   */
+  @Override
+  public void cleanupSplitsDir() throws IOException {
+    //do nothing, the split dir is the store dir itself, so we cannot delete it.
+  }
+
+  /**
+   * Do nothing. Here the merge dir is the table dir itself, we cannot delete it.
+   * @throws IOException
+   */
+  @Override
+  public void cleanupMergesDir() throws IOException {
+    //do nothing, the merges dir is the store dir itself, so we cannot delete it.
+  }
```

Comment:
What happens if a split or merge operation fundamentally fails (e.g. we fail to write the reference files in a split daughter region)? Does something in the SplitTableRegionProcedure remove the daughter region in that case?

Reply:
We have for merges, but not for splits. In the original implementation, these cleanup methods are only for temp directories. They are always called on region opening, so in the context of the direct store strategy, it was causing problems to me. Initial thought was to implement a cleaning solution on HBASE-25398. 

Or we can put a logic that checks if the dir is in StoreFilePathAccessor, deleting it otherwise.

Reply:
On a second thought about this, it doesn't help implement anything here. As mentioned above, the split cleanups only get called at region opening, which is fine for the default strategy, that will find rubbish under a valid, opening region dir ".split" dir. But in the context of direct store, those cleanup would only be called for splits that succeeded, which is not what we are seeking. And adding an additional check that queries system table every region open could be terrible.